core/psh/nc.c

```diff
@@ -0,0 +1,353 @@
+/*
+ * Phoenix-RTOS
+ *
+ * nc - TCP and UDP connect or listen
+ *
+ * Copyright 2021 Phoenix Systems
+ * Author: Maciej Purski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <arpa/inet.h>
+
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <sys/select.h>
+
+#include "psh.h"
+
+
+void psh_ncinfo(void)
+{
+	printf("  nc      - TCP and UDP connections and listens\n");
+}
+
+
+static void psh_nc_help(void)
+{
+	printf("Usage: nc [options] [destination] [port]\n");
+	printf("Options\n");
+	printf("  -4:  IPv4 only\n");
+	printf("  -6:  IPv6 only\n");
+	printf("  -h:  prints help\n");
+	printf("  -l:  listen mode, required port number\n");
+	printf("  -p:  source port\n");
+	printf("  -s:  source addr\n");
+	printf("  -u:  datagram socket\n");
+}
+
+
+/*
+ * This could be achieved using getaddrinfo(),
+ * however it does not recognize scope id for now
+ * and requires dynamic memory, which we
+ * are trying to avoid.
+ */
+static int psh_nc_sockaddrFill(int *af_family, char *addrstr, char *portstr,
+                               struct sockaddr_storage *sockaddr, socklen_t *addrlen)
+{
+	struct sockaddr_in6 *saddr_in6;
+	struct sockaddr_in *saddr;
+	char *ifname, *endptr;
+	int port = 0, scopeid = 0;
+	uint32_t addr[4] = { 0 };
+
+	if (*af_family == AF_UNSPEC) {
+		if (addrstr && strchr(addrstr, ':') != NULL)
+			*af_family = AF_INET6;
+		else
+			*af_family = AF_INET;
+	}
+
+	/* Try to resolve scope id */
+	if (*af_family == AF_INET6) {
+		if (addrstr == NULL)
+			return -1;
+		ifname = strtok(addrstr, "%");
+		if ((ifname = strtok(NULL, "%")) != NULL) {
+			scopeid = strtoul(ifname, &endptr, 10);
+			if (*endptr == '\n') {
```

Comment:
Newline character on command line?
When converting `portstr` you check `*endptr != '\0''` and here you don't?

Reply:
A stupid mistake... Thanks, I'll refactor this function.